### DISCUSSION TITLE
LRU parquet cache

### DISCUSSION THREAD

**Discussion by soerenwolfers** [duckdb/duckdb#14037] [Ideas]
I'd like a "parquet cache" that is updated with the bits of parquet that I last read as I go along exploring a big dataset

**Use case:** I have big (10TB) dataset stored in one or multiple parquet files. To explore this data, I often first look at small slices of it, e.g., a single day, making sanity checks and generating charts, before running queries on the full dataset.

Duckdb facilitates this quick interactive exploration via filter & projection pushdowns, which means that when I write, e.g., 

```sql
SELECT agent, sale
FROM read_parquet('file')
WHERE day = '2020-01-01'
```
then only the relevant data is read from disk into memory. 

However, if I spend a couple of hours investigating that day and I keep running queries like the above, then *each* of those queries reads from disk anew, which is slow (because even a single day contains 100GB of data).

To avoid this, I can load the specific day (or a subset of columns of that day) into a table via 
```sql
CREATE TABLE df AS (FROM read_parquet('file') WHERE day = '2020-01-01')
```
first, and then run my queries against this table. However, that's tedious in practice, where I regularly change the day/slice (e.g., a particular customer / sales region / ....) I'm drilling into as I find out new things / have to answer new questions and I don't want to manually create / drop / rename tables or add / remove columns after every query I run, i.e., manually manage what's essentially a LRU cache.

It would therefore be convenient if there was something like a "parquet cache" mechanism that I imagine could work as follows:

```sql
SELECT 
sum(sale)
FROM read_parquet('file', cache:=true) --- reads from disk on first execution and populates a behind-the-scenes-cache with the rows&columns of the parquet that are actually being read
WHERE day = '2020-01-01'
```

```sql
SELECT
median(sale)
FROM read_parquet('file', cache:=true) --- doesn't need to touch the disk at all
WHERE day = '2020-01-01'
```

```sql
SELECT
mean(sale)
FROM read_parquet('file', cache:=true) --- only reads second day from disk
WHERE day in ('2020-01-01', '2020-01-02')
```

```sql
SELECT
mean(sale)
FROM read_parquet('file', cache:=true) --- reads all other days and stores as much of it in the cache as is allowed by some configurable limit similar to/derived from the existing memory usage limits
```

The cache could be stored in a built-in table, that could be reset or inspected as needed. Not clear at all to me however what additional metadata beyond the actual row contents would need to be stored in the vachez however, to make sure a follow up scan knows whats needed from disk and what isn't.

**Comment by arjenpdevries** [comment]
So you propose that the query optimizer should know more about disk I/O through UDFs like `read_parquet`? You'd also want to cache a column from `read_csv`, not? 

(Did you check? I would expect DuckDB to already cache a column you reuse?)

**Reply by soerenwolfers** [reply]
anybody who uses csv, had already given up on fast IO ;) I found https://github.com/duckdb/duckdb/discussions/6948 saying there is no content cache, but there is a metadata cache that can be enabled via enable_object_cache
